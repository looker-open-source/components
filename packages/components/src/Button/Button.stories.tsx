/*

 MIT License

 Copyright (c) 2021 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */

import { AddCircle, Delete } from '@styled-icons/material'
import type { Page } from 'puppeteer'
import React from 'react'
import type { Story } from '@storybook/react/types-6-0'
import { defaultArgTypes as argTypes } from '../../../../apps/storybook/src/defaultArgTypes'
import { Button } from './Button'
import type { ButtonProps } from './types'

const Template: Story<ButtonProps> = args => <Button {...args} />

export const Basic = Template.bind({})
Basic.args = {
  children: 'Button Text',
}

export const Focused = Template.bind({})
Focused.args = {
  ...Basic.args,
  margin: 'small',
}

const beforeScreenshot = async (page: Page) => {
  const button = await page.$('button')
  await button?.type(' ')
  await page.waitForTimeout(50)
}
Focused.parameters = {
  beforeScreenshot,
  docs: { disable: true },
}

export const Critical = Template.bind({})
Critical.args = {
  ...Basic.args,
  color: 'critical',
}

export const CriticalFocused = Template.bind({})
CriticalFocused.args = {
  ...Critical.args,
  margin: 'small',
}

CriticalFocused.parameters = {
  beforeScreenshot,
}

export const Neutral = Template.bind({})
Neutral.args = {
  ...Basic.args,
  color: 'neutral',
}

export const NeutralFocused = Template.bind({})
NeutralFocused.args = {
  ...Neutral.args,
  margin: 'small',
}

NeutralFocused.parameters = {
  beforeScreenshot,
}

export const Disabled = Template.bind({})
Disabled.args = {
  ...Basic.args,
  disabled: true,
}

export const IconBefore = Template.bind({})
IconBefore.args = {
  ...Basic.args,
  iconBefore: <AddCircle />,
}

export const IconAfter = Template.bind({})
IconAfter.args = {
  ...Basic.args,
  iconAfter: <Delete />,
}

export const FullWidth = Template.bind({})
FullWidth.args = {
  ...Basic.args,
  fullWidth: true,
}

export const XSmall = Template.bind({})
XSmall.args = {
  ...Basic.args,
  size: 'xsmall',
}

export const Small = Template.bind({})
Small.args = {
  ...Basic.args,
  size: 'small',
}

export const Large = Template.bind({})
Large.args = {
  ...Basic.args,
  size: 'large',
}

export default {
  argTypes,
  component: Button,
  title: 'Button',
}
