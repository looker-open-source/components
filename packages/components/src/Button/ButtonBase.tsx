/*

 MIT License

 Copyright (c) 2019 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */

import {
  CompatibleHTMLProps,
  radii,
  RadiusSizes,
  reset,
  SemanticColors,
  space,
  SpaceProps,
} from '@looker/design-tokens'
import { rgba } from 'polished'
import React, { forwardRef, Ref, useState } from 'react'
import styled, { css } from 'styled-components'
import {
  minWidth,
  MinWidthProps,
  maxWidth,
  MaxWidthProps,
  WidthProps,
  width,
} from 'styled-system'
import { buttonSize, ButtonSizes, ButtonSizeProps } from './size'
import { ButtonIcon, buttonIcon, ButtonIconProps } from './icon'

export interface CustomizableButtonAttributes {
  borderRadius: RadiusSizes
}

export const CustomizableButtonAttributes: CustomizableButtonAttributes = {
  borderRadius: 'medium',
}

type ButtonColors = keyof SemanticColors

export interface ButtonBaseProps
  extends Omit<CompatibleHTMLProps<HTMLButtonElement>, 'type'>,
    ButtonSizeProps,
    MaxWidthProps,
    MinWidthProps,
    WidthProps,
    SpaceProps {
  type?: 'button' | 'submit' | 'reset'

  /**
   * Used along with `href` for an anchor link styled as a Button
   */
  forwardedAs?: 'a'

  /**
   * Defines the color of the button. Can be the string name of a color listed in the color theme, or a color object.
   * @default "primary"
   */
  color?: ButtonColors

  focusVisible?: boolean
}

export interface ButtonProps extends ButtonBaseProps, ButtonIconProps {
  size?: ButtonSizes
  /**
   * If true, the button's width will be set to 100%.
   */
  fullWidth?: boolean
}

export const buttonCSS = css<ButtonBaseProps>`
  ${reset}
  ${maxWidth}
  ${minWidth}
  ${width}

  ${props =>
    props.focusVisible &&
    `
    box-shadow: 0 0 0 0.15rem
      ${rgba(
        props.theme.colors.semanticColors[props.color || 'primary'].main,
        0.25
      )};
  `}

  align-items: center;
  border-radius: ${radii[CustomizableButtonAttributes.borderRadius]};
  cursor: pointer;
  display: inline-flex;
  font-weight: 600;
  justify-content: center;
  line-height: 1;
  outline: none;
  transition: border 80ms;
  vertical-align: middle;
  white-space: nowrap;

  &[disabled] {
    cursor: default;
    filter: grayscale(0.3);
    opacity: 0.25;
  }

  ${buttonSize}
  ${space}
`

const ButtonOuter = styled.button<ButtonProps>`
  ${buttonCSS}
  ${props => props.fullWidth && `width: 100%;`}
`

const ButtonJSX = forwardRef(
  (props: ButtonProps, ref: Ref<HTMLButtonElement>) => {
    const {
      children,
      iconBefore,
      iconAfter,
      onBlur,
      onKeyDown,
      ...restProps
    } = props

    const [isFocusVisible, setFocusVisible] = useState(false)

    const handleOnKeyUp = (event: React.KeyboardEvent<HTMLButtonElement>) => {
      setFocusVisible(true)
      onKeyDown && onKeyDown(event)
    }

    const handleOnBlur = (event: React.FocusEvent<HTMLButtonElement>) => {
      setFocusVisible(false)
      onBlur && onBlur(event)
    }

    return (
      <ButtonOuter
        focusVisible={isFocusVisible}
        onKeyUp={handleOnKeyUp}
        onBlur={handleOnBlur}
        {...restProps}
        ref={ref}
      >
        {iconBefore && <ButtonIcon name={iconBefore} />}
        {children}
        {iconAfter && <ButtonIcon name={iconAfter} />}
      </ButtonOuter>
    )
  }
)

ButtonJSX.displayName = 'ButtonJSX'

export const ButtonBase = styled(ButtonJSX)<ButtonProps>`
  ${buttonIcon}
`

ButtonBase.defaultProps = { size: 'medium' }
