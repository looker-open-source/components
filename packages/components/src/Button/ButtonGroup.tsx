/*

 MIT License

 Copyright (c) 2020 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */

import uniqueId from 'lodash/uniqueId'
import xor from 'lodash/xor'
import React, { ChangeEvent, FC, forwardRef, Ref } from 'react'
import styled from 'styled-components'
import { useControlWarn } from '../utils'
import { ButtonItem, ButtonItemLabel } from './ButtonItem'
import { ButtonGroupOrToggleProps, ButtonSet } from './ButtonSet'

const ButtonGroupFactory = forwardRef(
  (
    { onChange, value, ...props }: ButtonGroupOrToggleProps<string[]>,
    ref: Ref<HTMLDivElement>
  ) => {
    const isControlled = useControlWarn({
      controllingProps: ['onChange', 'value'],
      isControlledCheck: () => onChange !== undefined,
      name: 'ButtonGroup',
    })

    function handleChange(e: ChangeEvent<HTMLInputElement>) {
      const newValue = xor(value, [e.target.value])
      onChange && onChange(newValue)
    }

    return (
      <ButtonSet
        {...props}
        ref={ref}
        {...(isControlled
          ? {
              onChange: handleChange,
              value,
            }
          : { name: props.name || uniqueId() })}
      />
    )
  }
)

export const ButtonGroup = styled<FC<ButtonGroupOrToggleProps<string[]>>>(
  ButtonGroupFactory
)`
  ${ButtonItem} {
    height: 36px;
    border-radius: 4px;

    + ${ButtonItem} {
      margin-left: ${(props) => props.theme.space.xxsmall};
    }
  }

  ${ButtonItemLabel} {
    border-style: solid;
    border-width: 1px;
  }
`
