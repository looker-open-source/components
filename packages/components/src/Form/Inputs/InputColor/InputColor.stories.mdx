import { InputColor } from '@looker/components'

<Meta component={InputColor} title="Docs/Components/Inputs/InputColor" />

# InputColor

The `InputColor` component has an input to add a color value and a color wheel to pick color directly. The value prop can be used as a control component. To set an initial value on an uncontrolled component use defaultValue.

```js
<SpaceVertical>
  <InputColor />
  <InputColor value="green" />
  <InputColor defaultValue="purple" />
  <InputColor disabled />
  <InputColor defaultValue="green" disabled />
  <InputColor defaultValue="green" readOnly />
</SpaceVertical>
```
