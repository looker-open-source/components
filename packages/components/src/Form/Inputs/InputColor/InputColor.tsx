/*

 MIT License

 Copyright (c) 2020 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */

import React, {
  useState,
  ChangeEvent,
  FormEvent,
  forwardRef,
  Ref,
  useEffect,
} from 'react'
import styled from 'styled-components'
import get from 'lodash/get'
import { useID, useWrapEvent } from '../../../utils'
import { usePopover, PopoverContent } from '../../../Popover'
import { InputText, InputTextProps } from '../InputText'
import { useFormContext } from '../../Form'
import { HueSaturation, SimpleHSV } from './ColorWheel/color_wheel_utils'
import { ColorWheel } from './ColorWheel'
import { LuminositySlider } from './LuminositySlider'
import { Swatch } from './Swatch'
import {
  hsv2hex,
  simpleHSVtoFormattedColorString,
  str2simpleHsv,
} from './utils/color_format_utils'
import { isValidColor } from './utils/color_utils'

const colorWheelSize = 164

export interface InputColorProps extends Omit<InputTextProps, 'height'> {
  /**
   * If true, hides input and only show color swatch.
   */
  hideInput?: boolean
  value?: string
  defaultValue?: string
}

const createEventWithHSVValue = (
  color: SimpleHSV | string,
  name?: string
): ChangeEvent<HTMLInputElement> => {
  return {
    currentTarget: {
      name,
      value:
        typeof color === 'string'
          ? color
          : simpleHSVtoFormattedColorString(color),
    },
    target: {
      name,
      value:
        typeof color === 'string'
          ? color
          : simpleHSVtoFormattedColorString(color),
    },
  } as ChangeEvent<HTMLInputElement>
}

function getColorFromText(text?: string) {
  return text && isValidColor(text) ? str2simpleHsv(text) : undefined
}

export const InputColorComponent = forwardRef(
  (
    {
      className,
      hideInput,
      id,
      onChange,
      onFocus,
      onBlur,
      value,
      defaultValue,
      disabled,
      readOnly,
      ...props
    }: InputColorProps,
    ref: Ref<HTMLInputElement>
  ) => {
    const inputID = useID(id)
    const validationMessage = useFormContext(props)
    const initialColor = getColorFromText(value || defaultValue)

    const [color, setColor] = useState<SimpleHSV | undefined>(initialColor)
    const [inputTextValue, setInputTextValue] = useState(
      value || defaultValue || ''
    )
    const [isFocused, setIsFocused] = useState(false)

    const handleFocus = () => setIsFocused(true)
    const handleBlur = () => setIsFocused(false)
    const wrappedOnFocus = useWrapEvent(handleFocus, onFocus)
    const wrappedOnBlur = useWrapEvent(handleBlur, onBlur)

    useEffect(() => {
      if (value && value !== inputTextValue) {
        setColor(str2simpleHsv(value))
        !isFocused && setInputTextValue(value)
      }
    }, [isFocused, value, inputTextValue])

    const callOnChange = (newColor?: SimpleHSV | string) => {
      if (!onChange || !newColor) return
      onChange(createEventWithHSVValue(newColor, props.name))
    }

    const setColorState = (newColor: SimpleHSV) => {
      setColor(newColor)
      newColor && setInputTextValue(simpleHSVtoFormattedColorString(newColor))
      callOnChange(newColor)
    }

    const handleColorChange = (hs: HueSaturation) => {
      setColorState({ ...hs, v: get(color, 'v', 1) })
    }

    const handleSliderChange = (event: FormEvent<HTMLInputElement>) =>
      setColorState({
        ...(color || { h: 0, s: 100, v: 100 }),
        v: Number(event.currentTarget.value) / 100,
      })

    const handleInputTextChange = (event: FormEvent<HTMLInputElement>) => {
      const newValue = event.currentTarget.value
      setInputTextValue(newValue)

      const isValid = isValidColor(newValue)
      callOnChange(isValid ? newValue : undefined)
      setColor(getColorFromText(event.currentTarget.value))
    }

    const content = (
      <PopoverContent display="flex" flexDirection="column">
        <ColorWheel
          size={colorWheelSize}
          hue={get(color, 'h')}
          saturation={get(color, 's')}
          value={get(color, 'v')}
          onColorChange={handleColorChange}
        />
        <LuminositySlider
          min={0}
          max={100}
          step={1}
          value={get(color, 'v', 1) * 100}
          width={colorWheelSize}
          onChange={handleSliderChange}
        />
      </PopoverContent>
    )

    const { popover, domProps } = usePopover({ content })

    return (
      <div className={className}>
        <Swatch
          color={color ? hsv2hex(color) : undefined}
          disabled={disabled}
          readOnly={readOnly}
          {...domProps}
        />
        {!disabled && !readOnly && popover}
        {!hideInput && (
          <InputText
            {...props}
            aria-describedby={`${id}-describedby`}
            id={inputID}
            ref={ref}
            disabled={disabled}
            readOnly={readOnly}
            validationType={validationMessage && validationMessage.type}
            onChange={handleInputTextChange}
            value={inputTextValue}
            onFocus={wrappedOnFocus}
            onBlur={wrappedOnBlur}
          />
        )}
      </div>
    )
  }
)

InputColorComponent.displayName = 'InputColorComponent'

export const InputColor = styled(InputColorComponent)`
  display: flex;

  ${Swatch} {
    border-radius: ${({ hideInput, theme: { radii } }) =>
      hideInput ? radii.medium : radii.none};
    border-bottom-left-radius: ${({ theme: { radii } }) => radii.medium};
    border-right: ${({ hideInput }) => (hideInput ? undefined : 'none')};
    border-top-left-radius: ${({ theme: { radii } }) => radii.medium};
  }

  ${InputText} {
    border-radius: ${({ theme: { radii } }) => radii.none};
    border-bottom-right-radius: ${({ theme: { radii } }) => radii.medium};
    border-top-right-radius: ${({ theme: { radii } }) => radii.medium};
  }
`
