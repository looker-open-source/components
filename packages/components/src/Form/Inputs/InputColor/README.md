# InputColor

The `InputColor` component has an input to add a color value and a color wheel to pick color directly.

```tsx
function Basic(props: InputColorProps) {
  return <InputColor {...props} />;
}
```

## Value

Use the `value` prop for a controlled component.

```tsx
function ControlledColor() {
  const [color, setColor] = useState('red');

  function handleChange(value: string) {
    setColor(value);
  }

  function handleColorChange(e: React.ChangeEvent<HTMLInputElement>) {
    setColor(e.target.value);
  }
  return (
    <Space>
      <Select
        options={[{ value: 'green' }, { value: 'purple' }, { value: 'red' }]}
        value={color}
        onChange={handleChange}
        autoResize
      />
      <InputColor value={color} onChange={handleColorChange} />
      <Text>{color}</Text>
    </Space>
  );
}
```

## DefaultValue

To set an initial value on an uncontrolled component use `defaultValue`.

```tsx
function DefaultValue() {
  return <InputColor defaultValue="purple" />;
}
```

## Disabled

```tsx
function Disabled() {
  return <InputColor disabled value="purple" />;
}
```

```tsx
function DisabledNoValue() {
  return <InputColor disabled />;
}
```

## ReadOnly

```tsx
function ReadOnly() {
  return <InputColor readOnly value="purple" />;
}
```

## Swatch

```tsx
<Flex justifyContent="space-between">
  <Swatch />
  <Swatch color="rgb(79, 42, 186)" />
  <Swatch color="#FFFF00" />
</Flex>
```
