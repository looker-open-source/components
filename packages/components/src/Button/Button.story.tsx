/*

 MIT License

 Copyright (c) 2020 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */

import React from 'react'
import { Story } from '@storybook/react/types-6-0'
import { Button } from './Button'
import { ButtonProps } from './types'

const Template: Story<ButtonProps> = (args) => <Button {...args} />

export const Basic = Template.bind({})
Basic.args = {
  children: 'Button Text',
}

export const Critical = Template.bind({})
Critical.args = {
  ...Basic.args,
  color: 'critical',
}

export const Neutral = Template.bind({})
Neutral.args = {
  ...Basic.args,
  color: 'neutral',
}

export const Disabled = Template.bind({})
Disabled.args = {
  ...Basic.args,
  disabled: true,
}

export const IconBefore = Template.bind({})
IconBefore.args = {
  ...Basic.args,
  iconBefore: 'CircleAdd',
}

export const IconAfter = Template.bind({})
IconAfter.args = {
  ...Basic.args,
  iconAfter: 'Trash',
}

export const FullWidth = Template.bind({})
FullWidth.args = {
  ...Basic.args,
  fullWidth: true,
}

export const XSmall = Template.bind({})
XSmall.args = {
  ...Basic.args,
  size: 'xsmall',
}

export const Small = Template.bind({})
Small.args = {
  ...Basic.args,
  size: 'small',
}

export const Large = Template.bind({})
Large.args = {
  ...Basic.args,
  size: 'large',
}

export default {
  component: Button,
  title: 'Button',
}
