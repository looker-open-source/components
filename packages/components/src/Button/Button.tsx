/*

 MIT License

 Copyright (c) 2020 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */

import styled from 'styled-components'
import { ButtonBase } from './ButtonBase'

export const Button = styled(ButtonBase)`
  background: ${({ theme, color = 'key' }) => theme.colors[color]};
  border: 1px solid ${({ theme, color = 'key' }) => theme.colors[color]};
  color: ${({ theme, color = 'key' }) => theme.colors[`${color}Text`]};

  &:hover,
  &:focus,
  &.hover {
    background: ${({ theme, color = 'key' }) =>
      theme.colors[`${color}Interactive`]};
    border-color: ${({ theme, color = 'key' }) =>
      theme.colors[`${color}Interactive`]};
  }

  &[aria-expanded='true'],
  &:active,
  &.active {
    background: ${({ theme, color = 'key' }) =>
      theme.colors[`${color}Pressed`]};
    border-color: ${({ theme, color = 'key' }) =>
      theme.colors[`${color}Pressed`]};
  }

  &[disabled] {
    &:hover,
    &:active,
    &:focus {
      background-color: ${({ theme, color = 'key' }) => theme.colors[color]};
      border-color: ${({ theme }) => theme.colors.ui3};
    }
  }
`
