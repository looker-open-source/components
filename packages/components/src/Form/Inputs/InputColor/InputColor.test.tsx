/*

 MIT License

 Copyright (c) 2021 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */

import 'jest-styled-components'
import '@testing-library/jest-dom/extend-expect'
import React, { FormEvent, useState } from 'react'
import { fireEvent, screen } from '@testing-library/react'
import { renderWithTheme } from '@looker/components-test-utils'

import { Button } from '../../../Button'
import { InputColor } from './InputColor'

describe('InputColor', () => {
  test('with hidden input', () => {
    renderWithTheme(<InputColor value="yellow" hideInput />)
    expect(screen.queryByDisplayValue('yellow')).not.toBeInTheDocument()
  })

  test('starts with a named color value', () => {
    renderWithTheme(<InputColor value="green" />)
    expect(screen.getByDisplayValue('green')).toBeInTheDocument()
  })

  test('responds to input value change', () => {
    renderWithTheme(<InputColor value="green" />)
    const input = screen.getByDisplayValue('green')
    input.focus()
    fireEvent.change(input, { target: { value: 'blue' } })
    expect(screen.getByDisplayValue('blue')).toBeInTheDocument()
  })

  test('with controlled state', () => {
    function Wrapper() {
      const [value, setValue] = useState('')
      function handleClick() {
        setValue('yellow')
      }
      function handleChange(e: FormEvent<HTMLInputElement>) {
        setValue(e.currentTarget.value)
      }
      return (
        <>
          <Button onClick={handleClick}>Turn yellow</Button>
          <InputColor
            value={value}
            onChange={handleChange}
            placeholder="Select a color"
          />
        </>
      )
    }
    renderWithTheme(<Wrapper />)

    const button = screen.getByText('Turn yellow')
    const input = screen.getByPlaceholderText('Select a color')
    expect(input).toHaveValue('')
    fireEvent.click(button)
    expect(input).toHaveValue('yellow')
    fireEvent.change(input, { target: { value: 'purple' } })
    expect(input).toHaveValue('purple')
  })

  test('opens on swatch click', () => {
    renderWithTheme(<InputColor value="green" />)
    fireEvent.click(screen.getByTestId('swatch'))
    expect(screen.getByTestId('color-picker')).toBeInTheDocument()
    fireEvent.click(document)
  })

  test('can receive focus and blur handlers', () => {
    const onBlur = jest.fn()
    const onFocus = jest.fn()
    renderWithTheme(
      <InputColor onBlur={onBlur} onFocus={onFocus} value="green" />
    )

    const input = screen.getByDisplayValue('green')

    input.focus()
    expect(input).toHaveFocus()
    expect(onFocus).toHaveBeenCalled()

    input.blur()
    expect(input).not.toHaveFocus()
    expect(onBlur).toHaveBeenCalled()
  })

  test('changes color on <ColorPicker/> click', () => {
    renderWithTheme(<InputColor placeholder="Select a color" />)
    const input = screen.getByPlaceholderText(
      'Select a color'
    ) as HTMLInputElement

    fireEvent.click(screen.getByTestId('swatch'))

    const lightSaturationPreview = screen.getByTestId(
      'light-saturation-preview'
    )

    fireEvent.mouseDown(lightSaturationPreview, { clientX: 0, clientY: 0 })

    expect(input.value).toBe('#ffffff')

    /**
     * Close popover to silence act() warning
     * Note: Having just one click event produces the act() warning,
     * but having two events gets around this.
     *
     * Something about firing a mouseDown event (as opposed to a click
     * event) within the popover leads to funky test behavior. We need to
     * fire a mouseDown here, though, since the LightSaturationPreview uses mouse down
     * to the start the color handle movement / dragging.
     */
    fireEvent.click(document)
    fireEvent.click(document)
  })

  // mouseMove doesn't seem to move mouse position; leaving as xtest for now
  xtest('changes color on <ColorPicker/> mouse drag', () => {
    renderWithTheme(<InputColor placeholder="Select a color" />)
    const input = screen.getByPlaceholderText(
      'Select a color'
    ) as HTMLInputElement

    fireEvent.click(screen.getByTestId('swatch'))

    const lightSaturationPreview = screen.getByTestId(
      'light-saturation-preview'
    )

    fireEvent.mouseDown(lightSaturationPreview)

    fireEvent.mouseMove(lightSaturationPreview, { clientX: 200, clientY: 0 })
    expect(input.value).toBe('#ff0000')

    fireEvent.mouseUp(lightSaturationPreview)

    /**
     * Close popover to silence act() warning
     * Note: Having just one click event produces the act() warning,
     * but having two events gets around this.
     *
     * Something about firing a mouseDown event (as opposed to a click
     * event) within the popover leads to funky test behavior. We need to
     * fire a mouseDown here, though, since the LightSaturationPreview uses mouse down
     * to the start the color handle movement / dragging.
     */
    fireEvent.click(document)
    fireEvent.click(document)
  })

  test('disabled', () => {
    renderWithTheme(<InputColor disabled value="green" />)

    // Find input, verify it's disabled
    expect(screen.getByRole('textbox')).toBeDisabled()

    // Find swatch, verify clicking doesn't open Popover
    fireEvent.click(screen.getByTestId('swatch'))
    expect(screen.queryByTestId('color-picker')).not.toBeInTheDocument()
  })

  test('readOnly', () => {
    renderWithTheme(<InputColor readOnly value="green" />)

    // Find input, verify it's readOnly
    expect(screen.getByRole('textbox')).toHaveAttribute('readonly')

    // Find swatch, verify clicking doesn't open Popover
    fireEvent.click(screen.getByTestId('swatch'))
    expect(screen.queryByTestId('color-picker')).not.toBeInTheDocument()
  })
})
