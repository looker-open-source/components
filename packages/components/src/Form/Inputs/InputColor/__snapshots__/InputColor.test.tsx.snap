// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputColor Default render 1`] = `ShallowWrapper {}`;
